File: app/src/main/java/com/github/mbarrben/moviedb/movies/MoviesListModule.java
package com.github.mbarrben.moviedb.movies;

import com.github.mbarrben.moviedb.di.DomainModule;
import com.github.mbarrben.moviedb.domain.movies.GetMovies;
import com.github.mbarrben.moviedb.domain.movies.MoviesPresenter;
import com.github.mbarrben.moviedb.domain.movies.MoviesPresenterImpl;
import dagger.Module;
import dagger.Provides;

@Module(
    complete = false,
    includes = {
        DomainModule.class
    },
    injects = {
        MoviesActivity.class, MoviesFragment.class
    })
public class MoviesListModule {

  @Provides MoviesPresenter provideMoviesPresenter(GetMovies getMovies) {
    return new MoviesPresenterImpl(getMovies);
  }
}

File: app/src/main/kotlin/com/github/mbarrben/moviedb/movies/di/MoviesListModule.kt
package com.github.mbarrben.moviedb.movies.di

import com.github.mbarrben.moviedb.domain.movies.GetMovies
import com.github.mbarrben.moviedb.domain.movies.PopularMovies
import com.github.mbarrben.moviedb.domain.movies.SearchMovies
import com.github.mbarrben.moviedb.model.MovieRepository
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import javax.inject.Named

@Module
class MoviesListModule {

  @Provides fun getMovies(popular: PopularMovies, search: SearchMovies): GetMovies = GetMovies(popular, search)

  @Provides fun popularMovies(
      repo: MovieRepository,
      @Named("IO") subscribe: Scheduler,
      @Named("UI") observe: Scheduler
  ): PopularMovies = PopularMovies(repo, subscribe, observe)

  @Provides fun searchMovies(
      repo: MovieRepository,
      @Named("IO") subscribe: Scheduler,
      @Named("UI") observe: Scheduler
  ): SearchMovies = SearchMovies(repo, subscribe, observe)

}
